nbs/tsdataset.ipynb

```diff
@@ -54,7 +54,7 @@
     "#| export\n",
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #92.            for file in files:
</pre><p>Reading the whole files just to get the unique id and last time will be very slow. I think we can use the parquet metadata to extract this, something like:</p>

<pre class="ql-syntax" spellcheck="false">import pyarrow as pa

meta = pa.parquet.read_metadata(file)
rg = meta.row_group(0)
col2pos = {rg.column(i).path_in_schema: i for i in range(rg.num_columns)}
uid = rg.column(col2pos[target_col]).statistics.min
last_time = rg.column(col2pos[time_col]).statistics.max
for rg in range(1, meta.num_row_groups):
    last_time = max(last_time, rg.column(col2pos[time_col]).statistics.max)
</pre><p>We can also take the temporal cols as an argument and making sure they're there (by checking they're in the col2pos dict), that'd also allow using only a subset of features from the file.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/neuralforecast/pull/1049/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Reply:
yes makes sense - I've updated this so the temporal cols should now be directly input into the fit method (as shown in the example in core.ipynb) and we're just using the metadata to extract the relevant information

Reply:
<p>Second thought on this. Since the id will just be repeated it'd be nice to put it in the filename instead, so that we don't repeat it for the length of the series. As a bonus, if we make the filename something like <code>unique_id=id_0</code> then it'd also work when loading from a hive partitioned dataset.</p><p> </p><p>So can you please change this to something like <code>uid = Path(file).name.split('=')[-1]</code>? </p>

Reply:
Thanks for all the suggestions! I missed this in the first commit but have added it as part of the most recent commit

Reply:
<p>I'm sorry to do this again but I think the best format would be one that is compatible with hive partitioning, that'd make it very easy to take an existing dataframe and partition it with pandas for example <code>df.sort_values('ds').to_parquet('some_dir', partition_cols=['unique_id'])</code> or a very big dataset with spark:</p>

<pre class="ql-syntax" spellcheck="false">(
  df
  .repartitionByRange(numPartitions,"unique_id").
  .sortWithinPartitions("ds").
  .write.parquet("my_series.parquet")
)
</pre><p>So can you please change the logic to take a list of directories (which will be named: [unique_id=id_0, unique_id=id_1, ...]) where each one will hold one or more parquet files? <code>pd.read_parquet('unique_id=id_0')</code> will read all the partitions, so we just need to update the logic to get the metadata, I think something like this would work:</p>

<pre class="ql-syntax" spellcheck="false">for serie in series:
    serie_path = Path(serie)
    if not serie_path.is_directory():
        raise ValueError(f'paths must be directories, {serie} is not.')
&nbsp;&nbsp;&nbsp;&nbsp;uid = serie_path.name  # or serie_path.name.split('=')[-1]
    total_rows = 0
&nbsp;&nbsp;&nbsp;&nbsp;for partition in serie_path.iterdir():
        # add rows from partition and check it has the columns
</pre>

Reply:
No problem - I've updated this to support the dataframe being split between multiple files. I've had to re-add the logic to sort the dataframe though as we are now impacted by the order that the files are read in iterdir()!

Reply:
I'm not sure how to generate several files for a single partition in pandas, but spark names them as `part-00000`, `part-00001`, etc. which are read in that order by pandas when providing the directory only (we could also provide a list of the sorted filenames to be sure). So I think we can keep the `df = pd.read_parquet(self.files_ds.files[idx], columns=self.temporal_cols)` to avoid sorting on every read, which would make the loading very expensive.

Also, just in case more files are written, can you please change the iterator to `dir_path.glob('*.parquet')`